### ISSUE TITLE
Use submodules for Boost

### ISSUE THREAD

**Issue by vivid-synth** [supercollider/supercollider#2301]
Instead of just pasting its files into "external_libraries", we can have one or more submodules, from:

https://github.com/boostorg

The "super-project" is [here](https://github.com/boostorg/boost) and it's filled with submodules of its individual components:

https://github.com/boostorg/boost/tree/master/libs

**Comment by scztt** [comment]
👍

**Comment by vivid-synth** [comment]
Ok so I've been working on this, and writing scripts to find identical versions of individual files, so that we can first switch to submodules without any change in behavior, then incrementally upgrade boost as desired.

But am I overthinking this? Should we just go to the newest versions of the boost modules and see if anything breaks?

It would be nice to know if we're running any custom modules that have SC-specific patches/fixes, but I don't know if that's the case, or even if we are if they're necessary anymore.

**Comment by telephon** [comment]
Btw. will we get difficulties with git + submodules like that we had with vim?

**Comment by vivid-synth** [comment]
@telephon if you're talking about the annoyance when switching to commits that are older than the switch to submodules (or branches off those commits), I'd guess unfortunately there would be the same annoyances this time.

The upside, though, is that we'd have an up-to-date version of Boost and when they release a change we have a clear and clean way to upgrade.

E.g. boost has had thousands of updates in the past 2 years, but we've only gotten a few of them, and we've had to get them by just copy-pasting them into our repo.

```
$ git log --since="2 years" --pretty=oneline external_libraries/boost | wc -l
11
```

Probably the right way to do the upgrade to minimize annoyance would be to merge the changes in at a specified time, where e.g. all the devs have been warned a few days beforehand and can plan for it.

**Comment by scztt** [comment]
@vivid-synth I think that's totally the correct approach, but I doubt if it's pragmatically necessary to produce a functionally identical before/after build. I've seen many many many boost upgrades over the years, and I've only ever seen compile time problems from it. Isolating JUST the source tree changes is pretty unlikely to isolate any interesting runtime differences.

Regarding when to make the change - I'd do it in a branch to get a functioning travis build and unit test coverage, and then just announce a date (and warn people a bunch of times) if everything appears okay. It'd be good if several / many developers tried it first, to make sure all the build gotchas are worked out (it'll save a month of questions about "I'm getting this build error related to boost....")

**Comment by telephon** [comment]
@vivid-synth ok, I see. I really wonder why such a standard functionality like submodules are not supported better in git. Or maybe I miss something I should know …

**Comment by bagong** [comment]
The effect of the transition can be softened a bit by renaming the folder containing boost. Maybe somebody knows a better solution...

**Comment by vivid-synth** [comment]
Note we'll need to include https://github.com/supercollider/supercollider/pull/2457 and any other custom modifications

**Comment by vivid-synth** [comment]
https://github.com/supercollider/supercollider/pull/2473 is another custom one we need to support

**Comment by bagong** [comment]
Here is a start for systematic collection:
https://github.com/supercollider/supercollider/projects/2#card-1643180
Ready to be turned into something more visible and formal ;)

**Comment by brianlheim** [comment]
After working on #2772 I'm not convinced that this would be worth the bloat it would add (~180 MB) when working locally. As Tim Blechmann pointed out in a [mailing list thread](http://new-supercollider-mailing-lists-forums-use-these.2681727.n2.nabble.com/RE-boost-submodules-td7630455.html), the bcp utility makes it pretty easy to update when we need to. Although I don't like that we've patched Boost directly to get MSYS on Windows working, since we have that patch, that's also a reason not to move to submodules.

**Comment by bagong** [comment]
@brianlheim , chances are it will even get easier in the future, as the process get's more transparent and both boost and modern build environments support each other more robustly. I was not able myself to get bcp produce the files required on a Mac. Maybe you could be so kind to document the steps somewhere. Maybe I just used a wrong command line.

**Comment by brianlheim** [comment]
@bagong I had to chmod +x it. Make sure you run it from the main folder (not external_libraries) and point it at the boost root folder.

I figured whoever left it as non-executable had a good reason for doing so, but it looks like most of the scripts in the repo already have executable status. Maybe this one should be too.

**Comment by bagong** [comment]
Yea, I did the chmod thing, it was running just not succeeding. Well, I'll go and figure. At least I know that that it can be made work ;)

**Comment by vivid-synth** [comment]
@brianlheim I'm in favor of closing this and just going the bcp route, as long as we're applying the patches with the script too (with ```patch```, so we can just run external_libraries/extract_boost.sh and get what we need)

**Comment by brianlheim** [comment]
Closed by #2787, then. I left the patch as a separate file to be applied. Future Boost updates may make the gcc-ICE patch unnecessary.
